easypermissions/src/main/java/pub/devrel/easypermissions/EasyPermissions.java

```diff
@@ -85,205 +80,220 @@ public static boolean hasPermissions(@NonNull Context context, @NonNull String..
      *
-     * @param object
-     *         Activity or Fragment requesting permissions. Should implement
-     *         {@link android.support.v4.app.ActivityCompat.OnRequestPermissionsResultCallback}
-     *         or {@code android.support.v13.app.FragmentCompat.OnRequestPermissionsResultCallback}
-     * @param rationale
-     *         a message explaining why the application needs this set of permissions, will be displayed if the user rejects the request the first
-     *         time.
-     * @param positiveButton
-     *         custom text for positive button
-     * @param negativeButton
-     *         custom text for negative button
-     * @param requestCode
-     *         request code to track this request, must be < 256.
-     * @param perms
-     *         a set of permissions to be requested.
+     * @param activity       Activity requesting permissions. Should implement {@link
+     *                       ActivityCompat.OnRequestPermissionsResultCallback} {@code
+     *                       android.support.v13.app.FragmentCompat.OnRequestPermissionsResultCallback}
+     * @param rationale      a message explaining why the application needs this set of permissions,
+     *                       will be displayed if the user rejects the request the first time.
+     * @param positiveButton custom text for positive button
+     * @param negativeButton custom text for negative button
+     * @param requestCode    request code to track this request, must be < 256.
+     * @param perms          a set of permissions to be requested.
+     * @see android.Manifest.permission
      */
-    @SuppressLint("NewApi")
-    public static void requestPermissions(@NonNull final Object object, @NonNull String rationale,
-                                          @StringRes int positiveButton, @StringRes int negativeButton,
-                                          final int requestCode, @NonNull final String... perms) {
-
-        checkCallingObjectSuitability(object);
-
-        // Determine if rationale should be shown (generally when the user has previously
-        // denied the request);
-        boolean shouldShowRationale = false;
-        for (String perm : perms) {
-            shouldShowRationale =
-                    shouldShowRationale || shouldShowRequestPermissionRationale(object, perm);
+    @SuppressWarnings("NewApi")
+    public static void requestPermissions(@NonNull Activity activity,
+                                          @NonNull String rationale,
+                                          @StringRes int positiveButton,
+                                          @StringRes int negativeButton,
+                                          int requestCode,
+                                          @NonNull String... perms) {
+        if (hasPermissions(activity, perms)) {
+            notifyAlreadyHasPermissions(activity, requestCode, perms);
+            return;
         }
```

Comment:
This is the new model you'll see everywhere: if we already have the permissions or `api level< M`, then we're done.

We also hope that the api consumer has their `implements PermissionCallbacks` or annotations in the same object they pass into `requestPermissions`. If they do, we notify them which provides nice api consistency.